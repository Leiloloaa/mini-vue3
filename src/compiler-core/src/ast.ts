import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers";

/*
 * @Author: Stone
 * @Date: 2022-04-24 19:41:18
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:42:07
 */
export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT,
    COMPOUND_EXPRESSION
}

export function createVNodeCall(context, tag, props, children) {
    context.helper(CREATE_ELEMENT_VNODE);

    return {
        type: NodeTypes.ELEMENT,
        tag,
        props,
        children,
    };
}
