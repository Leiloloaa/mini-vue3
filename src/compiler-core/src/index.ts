/*
 * @Author: Stone
 * @Date: 2022-04-28 10:29:52
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-28 10:29:52
 */
// 对外导出函数都是要基于 index 函数

export * from './compile';
