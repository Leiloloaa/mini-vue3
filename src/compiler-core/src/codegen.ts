import { isString } from "../../shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"

/*
 * @Author: Stone
 * @Date: 2022-04-26 14:57:37
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:40:49
 */
export function generate(ast) {
    // 实现功能的步骤
    // 1、先知道要达到的效果
    // 2、任务拆分实现
    // 3、优化提取代码
    const context = createCodegenContext()
    const { push } = context

    genFunctionPreamble(ast, context)

    const functionName = "render"
    const args = ["_ctx", "_cache"]
    const signature = args.join(", ")

    push(`function ${functionName}(${signature}){`)
    push("return ")
    genNode(ast.codegenNode, context)
    push("}")

    return {
        code: context.code
    }
}

function createCodegenContext() {
    const context = {
        code: "",
        push(source) {
            context.code += source
        },
        helper(key) {
            return `_${helperMapName[key]}`
        }
    }
    return context
}

function genNode(node: any, context) {
    // 这里之前只处理 text 之后还需要处理别的类型 使用一个 switch
    switch (node.type) {
        case NodeTypes.TEXT:
            genText(node, context)
            break;
        case NodeTypes.INTERPOLATION:
            genInterpolation(node, context)
            break
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context)
            break
        case NodeTypes.ELEMENT:
            genElement(node, context)
            break
        case NodeTypes.COMPOUND_EXPRESSION:
            genCompoundExpression(node, context)
            break
        default:
            break;
    }
    // const { push } = context
    // push(`'${node.content}'`)
}

function genFunctionPreamble(ast: any, context) {
    const { push } = context
    const VueBinging = "Vue"
    // const helpers = ["toDisplayString"] // 帮助函数 后期需要实现 修改写在一个 helper 里面
    const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}` // 别名 带下划线
    if (ast.helpers.length > 0) {
        push(`const { ${ast.helpers.map(aliasHelper).join(", ")}} = ${VueBinging}`)
    }
    push("\n")
    push("return ")
}

function genText(node: any, context: any) {
    const { push } = context
    push(`'${node.content}'`)
}

function genInterpolation(node: any, context: any) {
    const { push, helper } = context
    push(`${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content, context)
    push(")")
}

function genExpression(node: any, context: any) {
    const { push } = context
    push(`'${node.content}'`)
}

function genElement(node, context) {
    const { push, helper } = context
    const { tag, children, props } = node
    // console.log('children', children)
    //   [ { type: 3, content: 'h1,' },
    //     { type: 0, content: { type: 1, content: 'message' } }
    //   ]
    // push(`${helper(CREATE_ELEMENT_VNODE)}("${tag}"), null, "hi," + _toDisplayString(_ctx.message)`)
    // element 里面的 children 一个一个拼接 循环遍历
    // const child = children[0]
    push(`${helper(CREATE_ELEMENT_VNODE)}(`)
    // for (let i = 0; i < children.length; i++) {
    //     const child = children[i];
    //     genNode(child, context)
    // }
    genNodeList(genNullable([tag, props, children]), context)
    // genNode(children, context)
    push(")")
}

function genNodeList(nodes, context) {
    const { push } = context
    for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        if (isString(node)) {
            push(node)
        } else {
            genNode(node, context)
        }
        if (i < nodes.length - 1) {
            push(", ")
        }
    }
}

function genNullable(args) {
    return args.map((arg) => arg || "null")
}

function genCompoundExpression(node: any, context: any) {
    const { push } = context
    const { children } = node
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isString(child)) {
            push(child)
        } else {
            genNode(child, context)
        }
    }
}
