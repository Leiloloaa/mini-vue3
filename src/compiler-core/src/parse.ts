import { NodeTypes } from "./ast";

const enum TagType {
    Start,
    End
}

export function baseParse(content) {
    const context = createParserContent(content)
    return createRoot(parseChildren(context, []))
}

function parseChildren(context, ancestors) {
    const nodes: any = []
    while (!isEnd(context, ancestors)) {
        let node
        const s = context.source;
        if (s.startsWith('{{')) {
            node = parseInterpolation(context)
        } else if (s[0] === "<") {
            // 需要用正则表达判断
            // <div></div>
            // /^<[a-z]/i/
            if (/[a-z]/i.test(s[1])) {
                node = parseElement(context, ancestors);
            }
        }
        if (!node) {
            node = parseText(context);
        }

        nodes.push(node)
    }
    return nodes
}

function isEnd(context, ancestors) {
    // 1、source 有值的时候
    // 2、当遇到结束标签的时候
    const s = context.source;
    if (s.startsWith('</')) {
        for (let i = ancestors.length - 1; i >= 0; i--) {
            const tag = ancestors[i].tag;
            if (startsWithEndTagOpen(s, tag)) {
                return true;
            }
        }
    }
    return !s;
}

function startsWithEndTagOpen(source, tag) {
    // 以左括号开头才有意义 并且 还需要转换为小写比较
    return (
        source.startsWith("</") &&
        source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
    );
}

function parseText(context) {
    const endToken = ['{{', '</'] // 停止的条件 如果同时存在 那么这个 index 要尽量的靠左 去最小的
    let endIndex = context.source.length // 停止的索引
    for (let i = 0; i < endToken.length; i++) {
        const index = context.source.indexOf(endToken[i])
        if (index !== -1 && endIndex > index) {
            endIndex = index
        }
    }

    // 解析文本 之前是 从头截取到尾部 但真是的环境是文本后面会有其它类型的 element 所以要指明停止的位置
    const content = parseTextData(context, endIndex)

    console.log('content -------', content);

    return {
        type: NodeTypes.TEXT,
        content
    }
}

function parseTextData(context, length) {
    const content = context.source.slice(0, length)

    advanceBy(context, length)
    return content
}

function parseElement(context, ancestors) {
    // 解析标签
    const element: any = parseTag(context, TagType.Start)
    ancestors.push(element)
    // 获取完标签后 需要把内部的 元素保存起来 需要用递归的方式去遍历内部的 element
    element.children = parseChildren(context, ancestors)
    ancestors.pop()

    // 这里要判断一下 开始标签和结束标签是否是一致的 不能直接消费完就 return
    if (startsWithEndTagOpen(context.source, element.tag)) {
        parseTag(context, TagType.End)
    } else {
        throw new Error(`缺少结束标签:${element.tag}`)
    }

    return element
}

function parseTag(context, type) {
    // <div></div>
    // 匹配解析
    // 推进
    const match: any = /^<\/?([a-z]*)/i.exec(context.source)
    const tag = match[1]
    // 获取完后要推进
    advanceBy(context, match[0].length)
    advanceBy(context, 1);

    if (type === TagType.End) return
    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}

function parseInterpolation(context) {
    // {{message}}
    // 拿出来定义的好处就是 如果需要更改 改动会很小
    const openDelimiter = '{{'
    const closeDelimiter = '}}'

    // 我们要知道关闭的位置
    // indexOf 表示 检索 }} 从 2 开始
    const closeIndex = context.source.indexOf(
        closeDelimiter,
        openDelimiter.length
    )

    // 删除 前两个字符串
    // context.source = context.source.slice(openDelimiter.length)
    advanceBy(context, openDelimiter.length)

    // 内容的长度就等于 closeIndex - openDelimiter 的长度
    const rawContentLength = closeIndex - openDelimiter.length
    const rawContent = parseTextData(context, rawContentLength)
    const content = rawContent.trim()

    // 然后还需要把这个字符串给删了 模板是一个字符串 要接着遍历后面的内容
    // context.source = context.source.slice(rawContentLength + closeDelimiter.length);
    advanceBy(context, closeDelimiter.length)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content
        }
    }
}

function advanceBy(context, length) {
    context.source = context.source.slice(length);
}

function createRoot(children) {
    return { children, type: NodeTypes.ROOT }
}

function createParserContent(content) {
    return {
        source: content
    }
}
