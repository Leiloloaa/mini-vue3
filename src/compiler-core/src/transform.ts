/*
 * @Author: Stone
 * @Date: 2022-04-25 17:19:47
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:37:21
 */

import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

// options 提供了更动态的传参方式
export function transform(root, options = {}) {
    // 任务拆分
    // 1 遍历 - 深度优先遍历 和 广度优先遍历
    // 2 修改 test content

    // 创建上下文本
    const context = createTransformConetext(root, options)
    traverseNode(root, context)
    createRootCodegen(root)

    root.helpers = [...context.helpers.keys()]
    console.log('root.helpers', root.helpers)
}

function createTransformConetext(root: any, options: any): any {
    const context = {
        root,
        nodeTransforms: options.nodeTransforms || [], // 插件列表
        helpers: new Map(),
        helper(key) {
            context.helpers.set(key, 1)
        }
    }
    return context
}

function traverseNode(node: any, context) {
    const nodeTransforms = context.nodeTransforms
    const exitFns: any = []
    for (let i = 0; i < nodeTransforms.length; i++) {
        // 调用插件
        const transform = nodeTransforms[i];
        const onExit = transform(node, context)
        if (onExit) exitFns.push(onExit)
    }

    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break;
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(node, context)
            break;
        default:
            break;
    }
    let i = exitFns.length
    while (i--) {
        exitFns[i]()
    }
}

function traverseChildren(node: any, context: any) {
    const children = node.children
    for (let i = 0; i < children.length; i++) {
        const node = children[i];
        traverseNode(node, context)
    }
}

function createRootCodegen(root: any) {
    const child = root.children[0]
    if (child.type === NodeTypes.ELEMENT) {
        root.codegenNode = child.codegenNode
    } else {
        root.codegenNode = root.children[0]
    }
}