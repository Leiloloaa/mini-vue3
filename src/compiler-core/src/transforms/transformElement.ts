import { createVNodeCall, NodeTypes } from "../ast";

/*
 * @Author: Stone
 * @Date: 2022-04-27 15:23:00
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:43:06
 */
export function transformElement(node, context) {
    if (node.type === NodeTypes.ELEMENT) {

        return () => {
            // 中间处理层

            // tag
            const vnodeTag = `'${node.tag}'`

            // props
            let vnodeProps

            // children
            const { children } = node
            let vnodeChildren = children[0]
            node.codegenNode = createVNodeCall(
                context,
                vnodeTag,
                vnodeProps,
                vnodeChildren
            );
        }

    }
}
