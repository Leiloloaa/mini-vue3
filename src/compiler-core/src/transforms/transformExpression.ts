import { NodeTypes } from "../ast"

/*
 * @Author: Stone
 * @Date: 2022-04-27 14:52:31
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 15:02:45
 */
export function transformExpression(node) {
    if (node.type === NodeTypes.INTERPOLATION) {
        node.content = processExpression(node.content)
    }
}

function processExpression(node) {
    node.content = `_ctx.${node.content}`
    return node
}