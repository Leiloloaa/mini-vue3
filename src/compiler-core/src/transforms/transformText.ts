import { NodeTypes } from "../ast";
import { isText } from "../utils";

/*
 * @Author: Stone
 * @Date: 2022-04-27 15:58:33
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:41:50
 */
export function transformText(node) {
    // 实现一个 compose 类型的节点
    // 目的是将 element 类型下的 chilren + 起来(注意 是一个接一个的)
    if (node.type === NodeTypes.ELEMENT) {

        return () => {
            const { children } = node

            let currentContainer
            for (let i = 0; i < children.length; i++) {
                const child = children[i];
                if (isText(child)) {
                    for (let j = i + 1; j < children.length; j++) {
                        const next = children[j];
                        if (isText(next)) {
                            if (!currentContainer) {
                                currentContainer = children[i] = {
                                    type: NodeTypes.COMPOUND_EXPRESSION,
                                    children: [child]
                                }
                            }
                            currentContainer.children.push(" + ")
                            currentContainer.children.push(next)
                            children.splice(j, 1)
                            j--
                        } else {
                            currentContainer = undefined
                            break
                        }
                    }
                }
            }
        }
    }
}