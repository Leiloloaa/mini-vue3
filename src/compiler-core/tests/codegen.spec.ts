/*
 * @Author: Stone
 * @Date: 2022-04-26 14:57:19
 * @LastEditors: Stone
 * @LastEditTime: 2022-04-27 18:39:21
 */
import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from './../src/transforms/transformElement';
import { transformText } from './../src/transforms/transformText';

describe("codegen", () => {
    it("string", () => {
        const ast = baseParse("hi");
        transform(ast);
        const { code } = generate(ast);
        // 这是生成快照  如果需要更新快照 运行命令后面加上 -u
        expect(code).toMatchSnapshot();
    });

    it("interpolation", () => {
        const ast = baseParse("{{message}}");
        transform(ast, {
            nodeTransforms: [transformExpression]
        });
        const { code } = generate(ast);
        expect(code).toMatchSnapshot();
    });

    it("element", () => {
        const ast = baseParse("<div></div>");
        transform(ast, {
            nodeTransforms: [transformElement]
        });
        const { code } = generate(ast);
        expect(code).toMatchSnapshot();
    });

    it("union", () => {
        const ast:any = baseParse("<div>h1,{{message}}</div>");
        transform(ast, {
            nodeTransforms: [transformExpression, transformElement, transformText]
        });
        console.log('ast -----', ast);

        const { code } = generate(ast);
        expect(code).toMatchSnapshot();
    });
});