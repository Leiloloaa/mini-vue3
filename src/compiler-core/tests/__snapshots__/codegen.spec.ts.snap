// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementVnode:_createElementVnode} = Vue
return function render(_ctx, _cache){return _createElementVnode('div', null, null)}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString:_toDisplayString} = Vue
return function render(_ctx, _cache){return _toDisplayString('_ctx.message')}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache){return 'hi'}"`;

exports[`codegen union 1`] = `
"const { toDisplayString:_toDisplayString, createElementVNode:_createElementVNode} = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'h1,' + _toDisplayString('_ctx.message'))}"
`;
